text/0000-loop-break-value.md

```diff
@@ -0,0 +1,290 @@
+[motivation]: #motivation
+
+> Rust is an expression-oriented language. Currently loop constructs don't
+> provide any useful value as expressions, they are run only for their
+> side-effects. But there clearly is a "natural-looking", practical case,
+> described in [this thread](https://github.com/rust-lang/rfcs/issues/961)
+> and [this] RFC, where the loop expressions could have
+> meaningful values. I feel that not allowing that case runs against the
+> expression-oriented conciseness of Rust.
+> [comment by golddranks](https://github.com/rust-lang/rfcs/issues/961#issuecomment-220820787)
+
+Some examples which can be much more concisely written with this RFC:
+
+```rust
+// without loop-break-value:
+let x = {
+    let temp_bar;
+    loop {
+        ...
+        if ... {
+            temp_bar = bar;
+            break;
+        }
+    }
+    foo(temp_bar)
+};
+
+// with loop-break-value:
+let x = foo(loop {
+        ...
+        if ... { break bar; }
+    });
+
+// without loop-break-value:
+let computation = {
+    let result;
+    loop {
+        if let Some(r) = self.do_something() {
+            result = r;
+            break;
+        }
+    }
+    result.do_computation()
+};
+self.use(computation);
+
+// with loop-break-value:
+let computation = loop {
+        if let Some(r) = self.do_something() {
+            break r;
+        }
+    }.do_computation();
+self.use(computation);
+```
+
+# Detailed design
+[design]: #detailed-design
+
+This proposal does two things: let `break` take a value, and let `loop` have a
+result type other than `()`.
+
+### Break Syntax
+
+Four forms of `break` will be supported:
+
+1.  `break;`
+2.  `break 'label;`
+3.  `break EXPR;`
+4.  `break 'label EXPR;`
+
+where `'label` is the name of a loop and `EXPR` is an expression.
+
+### Result type of loop
+
+Currently the result type of a 'loop' without 'break' is `!` (never returns),
+which may be coerced to any type), and the result type of a 'loop' with 'break'
+is `()`. This is important since a loop may appear as
+the last expression of a function:
+
+```rust
+fn f() {
+    loop {
+        do_something();
+        // never breaks
+    }
+}
+fn g() -> () {
+    loop {
+        do_something();
+        if Q() { break; }
+    }
+}
+fn h() -> ! {
+    loop {
+        do_something();
+        // this loop is not allowed to break due to inferred `!` type
+    }
+}
+```
+
+This proposal changes the result type of 'loop' to `T`, where:
+
+*   if a loop is "broken" via `break;` or `break 'label;`, the loop's result type must be `()`
+*   if a loop is "broken" via `break EXPR;` or `break 'label EXPR;`, `EXPR` must evaluate to type `T`
+*   as a special case, if a loop is "broken" via `break EXPR;` or `break 'label EXPR;` where `EXPR` evaluates to type `!` (does not return), this does not place a constraint on the type of the loop
+*   if external constaint on the loop's result type exist (e.g. `let x: S = loop { ... };`), then `T` must be coercible to this type
```

Comment:
This list still has issues with handling of coercion. For example I imagine something like this ought to be valid:

```
loop {
     // ...
     break (expr: SomeT);
     // ...
     break (expr: OtherT);
}: RT
```

where `SomeT` and `OtherT` both coerce to `RT`. If *this* discrepancy was fixed, the list could avoid having to contain the last two notes/requirements. I also dislike those two requirements, because they have potential to diverge from type rules elsewhere in the language.

(Why this matters? Substitute `SomeT = &[u8; 42]`, `OtherT = &[u8; 32]`, `RT = &[u8]`.)